redbox-core/redbox/graph/nodes/processes.py

```diff
@@ -51,12 +51,19 @@ def build_retrieve_pattern(
 
 def build_chat_pattern(
     prompt_set: PromptSet,
+    tools: list[StructuredTool] | None = None,
     final_response_chain: bool = False,
 ) -> Runnable[RedboxState, dict[str, Any]]:
-    """Returns a Runnable that uses state["request"] to set state["text"]."""
+    """Returns a Runnable that uses the state to set state["text"].
+
+    If tools are supplied, can also set state["tool_calls"].
+    """
 
     def _chat(state: RedboxState) -> dict[str, Any]:
         llm = get_chat_llm(Settings(), state["request"].ai_settings)
+        if tools:
```

Comment:
thinking out loud again: we are repeating this pattern 3 times, is there an easier way? no need to do anything this can be re-visitied later